pex/pip.py

```diff
@@ -324,6 +340,179 @@ def analyze(self, line):
+                ),
+                via=resolved_requirement.via,
+            )
+
+    pin = attr.ib()  # type: Pin
+    artifact = attr.ib()  # type: PartialArtifact
+    requirement = attr.ib()  # type: Requirement
+    additional_artifacts = attr.ib(default=())  # type: Tuple[PartialArtifact, ...]
+    via = attr.ib(default=())  # type: Tuple[str, ...]
+
+    def _iter_urls_to_fingerprint(self):
+        # type: () -> Iterator[str]
+        if not self.artifact.fingerprint:
+            yield self.artifact.url
+        for artifact in self.additional_artifacts:
+            if not artifact.fingerprint:
+                yield artifact.url
+
+
+class Locker(_LogAnalyzer):
+    class StateError(Exception):
+        """Indicates the Locker lifecycle was violated.
+
+        A Locker 1st should be used to collect data from a Pip debug log and only then can a `lock`
+        be requested.
+        """
+
+    def __init__(
+        self,
+        target,  # type: DistributionTarget
+        lock_configuration,  # type: LockConfiguration
+        network_configuration=None,  # type: Optional[NetworkConfiguration]
+    ):
+        # type: (...) -> None
+        self._target = target
+        self._lock_configuration = lock_configuration
+        self._resolved_requirements = []  # type: List[ResolvedRequirement]
+        self._links = defaultdict(OrderedSet)  # type: DefaultDict[Pin, OrderedSet[PartialArtifact]]
+        self._url_fetcher = URLFetcher(
+            network_configuration=network_configuration, handle_file_urls=True
+        )
+        self._analysis_completed = False
+        self._locked_resolve = None  # type: Optional[LockedResolve]
+
+    def should_collect(self, returncode):
+        # type: (int) -> bool
+        return returncode == 0
+
+    @staticmethod
+    def _extract_resolve_data(url):
+        # type: (str) -> Tuple[Pin, PartialArtifact]
+
+        fingerprint = None  # type: Optional[Fingerprint]
+        fingerprint_match = re.search(r"(?P<url>[^#]+)#(?P<algorithm>[^=]+)=(?P<hash>.*)$", url)
+        if fingerprint_match:
+            url = fingerprint_match.group("url")
+            algorithm = fingerprint_match.group("algorithm")
+            hash_ = fingerprint_match.group("hash")
+            fingerprint = Fingerprint(algorithm=algorithm, hash=hash_)
+
+        pin = Pin.canonicalize(ProjectNameAndVersion.from_filename(urlparse.urlparse(url).path))
+        partial_artifact = PartialArtifact(url, fingerprint)
+        return pin, partial_artifact
+
+    def analyze(self, line):
+        # type: (str) -> _LogAnalyzer.Continue[None]
+
+        if self._analysis_completed:
+            raise self.StateError("Line analysis was requested after the log analysis completed.")
+
+        match = re.search(
+            r"Added (?P<requirement>.*) from (?P<url>[^\s]+) (\(from (?P<from>.*)\) )?to build "
+            r"tracker",
+            line,
+        )
+        if match:
```

Comment:
To check my understanding: if this parser does not capture a certain requirement—for whatever reason—the lock will be missing an entry. We don't validate at the end that the `LockedResolve` has entries for everything we expect. Instead, the lockfile will be missing an entry and at consumption time it will fail because `--hash` requires it to be fully comprehensive. Is that right?

> for everything we expect

We don't compute what to expect, right? Whatever pip gives back is what we roll with? (I know we can't anticipate  things like the --hash and version, but I'm wondering about project_name based on metadata of direct deps.)

Reply:
Your understanding is correct, if the vendored Pip has unexpected code paths (or else a randomizer?) that somehow outputs log lines we don't expect or omits ones we do expect, then this scheme will fail, but only at lock consumption time later when Pip --hash mode complains about missing --hashes. Hopefully the first sentence underscores that we'll never hit that Pip --hash checking backstop though, because this is a vendored and completely understood version of Pip. It is true that on upgrade of vendored Pip, the ITs will have to ensure the logging is still as we expect / the human upgrading the vendored Pip will need to take care - but that's already a given.